poller-service with all locking in Python

clinta opened the pull request:
This update to poller-service allows it to function with the php changes as reverted in #1836.

To facilitate all locking within python, poller-service now opens a unique mysql connection per thread. Since there are no php changes this should have no effect on anyone who is not runnning poller-service.

adammmmm commented:
Made a PR #1858 and tried this PR with help from f0o. Used these commands:

git remote add clinta https://github.com/clinta/librenms.git
git remote update
git checkout origin/master
git checkout -b PollerServiceTest
git merge clinta/poller-service2

I then started poller-service again, and get the same output in /var/log/syslog:

poller-service: INFO: Starting poller-service
poller-service: ERROR: MySQL query error; Is your schema up to date?

Let me know if there's any output you need!

clinta commented:
@adammmmm, what version of MySQL are you running?

adammmmm commented:
@clinta Server version: 5.5.44-0ubuntu0.14.04.1 (Ubuntu)

clinta commented:
I'm testing now. I'm thinking 408b8d may have fixed a bug in MariaDB, while braking it in MySQL.

adammmmm commented:
Alright! Just shout if you need me to test anything.

clinta commented:
Please run git pull and try starting the poller-service again.

adammmmm commented:
Did, seemed to have done the trick, great!

poller-service: INFO: Starting poller-service
poller-service: INFO: 80 devices scanned in the last minute

clinta commented:
@spinza, please give this branch a try when you get a chance. It is working well for me.

adammmmm commented:
Maybe I spoke too quickly, it works perfectly for a while but then all of a sudden it shuts down and respawns.
init: poller-service main process (6084) terminated with status 1
init: poller-service main process ended, respawning
poller-service: INFO: Starting poller-service
init: poller-service main process (2208) terminated with status 1
init: poller-service main process ended, respawning
poller-service: INFO: Starting poller-service
INFO: 107 devices scanned in the last minute

spinza commented:
@clinta Will try to get on this this weekend.

spinza commented:
The differences between mysql and mariadb are getting annoying.

clinta commented:
@adammmmm, please pull and give this a try now. I had my pollers stop functioning sometime last night after running solid for 12 hours. It looks like it would hang if one of the MySQL connections would drop at the right time.

Hard to tell if that's the same problem you're having or not, but I've added some more error logging in the MySQL connections and queries, so that should give a better clue.

Because I rebased to get the latest librenms updates, you may need to do this to pull the changes:
```bash
git fetch --all
git reset --hard clinta/poller-service2
```

adammmmm commented:
@clinta alright, done, I'll let it run over night

adammmmm commented:
Didn't have to run for too long, got it just now:

Sep  9 15:21:12 localhost kernel: [25826.100087] init: poller-service main process (19683) terminated with status 1
Sep  9 15:21:12 localhost kernel: [25826.100103] init: poller-service main process ended, respawning
Sep  9 15:22:01 localhost CRON[27393]: (librenms) CMD (   /opt/librenms/alerts.php >> /dev/null 2>&1)
Sep  9 15:22:12 localhost poller-service: INFO: Starting poller-service
Sep  9 15:22:12 localhost poller-service: WARNING: MySQL Error during query, reconnecting.
Sep  9 15:23:01 localhost CRON[6627]: (librenms) CMD (   /opt/librenms/alerts.php >> /dev/null 2>&1)
Sep  9 15:22:22 localhost poller-service: message repeated 16 times: [ WARNING: MySQL Error during query, reconnecting.]
Sep  9 15:23:15 localhost poller-service: INFO: 102 devices scanned in the last minute
Sep  9 15:24:01 localhost CRON[21471]: (librenms) CMD (   /opt/librenms/alerts.php >> /dev/null 2>&1)
Sep  9 15:24:15 localhost poller-service: INFO: 94 devices scanned in the last minute

clinta commented:
@adammmmm, what's the last entry in `/var/log/upstart/poller-service.log`, assuming you're using the upstart config.

adammmmm commented:
Traceback (most recent call last):
  File "/opt/librenms/poller-service.py", line 329, in <module>
    if not getThreadQueueLock(device_id):
  File "/opt/librenms/poller-service.py", line 204, in getThreadQueueLock
    if getLock('queue.{0}'.format(device_id), thread_cursor.db):
  File "/opt/librenms/poller-service.py", line 186, in getLock
    cursor = db.query(query)
  File "/opt/librenms/poller-service.py", line 65, in query
    cursor.execute(sql)
  File "/usr/lib/python2.7/dist-packages/MySQLdb/cursors.py", line 174, in execute
    self.errorhandler(self, exc, value)
  File "/usr/lib/python2.7/dist-packages/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorclass, errorvalue
_mysql_exceptions.ProgrammingError: (2014, "Commands out of sync; you can't run this command now")
MIB search path: $HOME/.snmp/mibs:/usr/share/snmp/mibs:/usr/share/snmp/mibs/iana:/usr/share/snmp/mibs/ietf:/usr/share/mibs/site:/usr/share/snmp/mibs:/usr/share/mibs/iana:/usr/share/mibs/ietf:/usr/share/mib
s/netsnmp:/opt/librenms/mibs
Cannot find module (MTA-MIB): At line 1 in (none)
Cannot find module (NETWORK-SERVICES-MIB): At line 1 in (none)
*** Error in `python': double free or corruption (!prev): 0x0000000002578c90 ***
MIB search path: $HOME/.snmp/mibs:/usr/share/snmp/mibs:/usr/share/snmp/mibs/iana:/usr/share/snmp/mibs/ietf:/usr/share/mibs/site:/usr/share/snmp/mibs:/usr/share/mibs/iana:/usr/share/mibs/ietf:/usr/share/mib
s/netsnmp:/opt/librenms/mibs
Cannot find module (MTA-MIB): At line 1 in (none)
Cannot find module (NETWORK-SERVICES-MIB): At line 1 in (none)
MIB search path: $HOME/.snmp/mibs:/usr/share/snmp/mibs:/usr/share/snmp/mibs/iana:/usr/share/snmp/mibs/ietf:/usr/share/mibs/site:/usr/share/snmp/mibs:/usr/share/mibs/iana:/usr/share/mibs/ietf:/usr/share/mib
s/netsnmp:/opt/librenms/mibs
Cannot find module (MTA-MIB): At line 1 in (none)
Cannot find module (NETWORK-SERVICES-MIB): At line 1 in (none)
Traceback (most recent call last):
  File "/opt/librenms/poller-service.py", line 346, in <module>
    if not getThreadActionLock(device_id, action):
  File "/opt/librenms/poller-service.py", line 219, in getThreadActionLock
    if getLock('{0}.{1}'.format(action, device_id), thread_cursor.db):
  File "/opt/librenms/poller-service.py", line 186, in getLock
    cursor = db.query(query)
  File "/opt/librenms/poller-service.py", line 65, in query
    cursor.execute(sql)
  File "/usr/lib/python2.7/dist-packages/MySQLdb/cursors.py", line 174, in execute
    self.errorhandler(self, exc, value)
  File "/usr/lib/python2.7/dist-packages/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorclass, errorvalue
_mysql_exceptions.InterfaceError: (0, 'Lost connection to MySQL server during query')

clinta commented:
@adammmmm, please try these latest updates, they've run stable all night for me.

adammmmm commented:
I've run 01bef1f since the morning which has run well so far. I'll Update to a761a4b and let it run through the night!

adammmmm commented:
It's now run through the night without any issue whatsoever, great work!

f0o commented:
@adammmmm please let it run over the weekend, if there are no issues by monday I will merge the PR.

Great work you two :)

adammmmm commented:
Will do, just glad to be of help to this great project without knowing how to code :)

spinza commented:
Activate now on my system

spinza commented:
@clinta this seems to have a large increase in mysql connections.  How many are expected to be open for 16 workers?

spinza commented:
I have 3 pollers with 16 workers each and I often have 140 connections to my database?  Seems excessive?  I have 2 other pollers that poll specific devices only, so they should only have a handful of workers up at any time.

adammmmm commented:
Just an update, still no crash, still working fine in my environment.

spinza commented:
I read the code now more clearly.  It launches 16 workers for each poller regardless of how how much work there is.  For me this means with 5 pollers you'd have 5 * 16 = 80 connections.  I can scale back the worker count to 1 on pollers that are only checking a small number of devices.  I'm seeing 140 connections appearing, so not sure what causes this.  Way more than 80?  It doesn't seem right.  I'm seeing some operational errors as well.  

Which seems to cause it to reconnect.  Is this perhaps not the problem?
Or is it a timeout?

adammmmm commented:
1 poller with 16 workers here. According to mysql status 'max_used_connections' mine's using 55 connections.

spinza commented:
Once a worker runs out of work it polls for work every second.  So 16 workers means 16 queries per second to the mysql server to poll for work.  Times 5 gives you 80 queries per second checking for work.

clinta commented:
Yes, this does have way more connections. The problem is that on any version of MySQL or MariaDB < 10.2 you can only hold one MySQL lock per connection. Since I can't rely on php making unique connections per poll as I origionally assumed, I need to make one connection per thread in python, which is in addition to the connections that come from php.

My pollers never run out of work so I hadn't thought much about the behavior you're seeing where you're getting a query per thread every second. I should be able to improve that situation pretty easily.

spinza commented on poller-service.py:
```diff
@@ -48,6 +49,40 @@
+                ret = cursor.fetchall()
+                cursor.close()
+                return ret
+            except (AttributeError, MySQLdb.OperationalError):
```
Any chance with this exception that the connection is still active?

spinza commented on poller-service.py:
```diff
@@ -211,96 +213,110 @@ def sleep_until(timestamp):
+        dev_row = db.query(dev_query)
+        if len(dev_row) < 1:
+            # Sleep 1 second after getting an empty query, don't hammer the sql server for no reason.
+            time.sleep(1)
```
Maybe this can be extended.  Sleep 5?

clinta commented:
@spinza, I don't believe so, but I'm not confident enough to say it's impossible. So I added a call to conn.close() before reconnecting.

spinza commented:
Sorry I was wondering why I'm seeing 140 connections (i.e. more than 80) and it was because of php also connecting.  Didn't think it through.  So in theory if the worker threads are all busy this could go up to 160 connections (i.e. twice the total worker threads).

I will probably not be able to handle this many connections.  I'm a little resource constrained.  So I guess I will change this to main thread work polling (one mysql connection) with memcached based locking.  I.e. similar to the old poller-wrapper.py except it runs as a service?  

Will you be open to generalise this to that sort of solution?  Or should I start a poller-service-memchached.py PR?  Could we do it such that memcached is an option flag? If you changed your code to main thread work fetching.  We can then either use memcached or mysql to do the "locking".

I guess this PR could be split into two parts.  Part (a) is change to service, (b) is to change to mysql based locking instead of memcached.

I'm keen on (a) but less on (b) for now.  Also part (b) has been holding part (a) back. I'm just going by my circumstances, not sure how others feel.  I may be on my own.

clinta commented:
Is running MariaDB 10.2 or newer an option for you? I've been thinking I'd like to add a configurable option to use a single mysql connection that would work with MariaDB 10.2, but I wasn't sure if anyone but me would ever use it.

clinta commented:
Also, give the latest comits a try. It will still make the same number of connections, but now the number of queries are way fewer, maximum of 1 per second and user configurable using $['poller_service_retry_query']. The sleeping connections may not have as big of a performance impact as you might think.

clinta commented:
This latest update adds support for using a single connection for MariaDB 10.2 or newer. Enable it by setting `$config['poller_service_single_connection'] = true;`.

In my case it has reduced the active MariaDB threads from around 115 to about 60.

f0o commented:
!trigger ci

f0o commented:
trigger ci

spinza commented:
Sorry I can't upgrade now.  

This is breaking it for me.  I actually have had to stop all my pollers.

adammmmm commented:
When I updated from cb3189e to 5e3d968 I saw an increase in context switches, interrupts, processor usage and load average, previously it's been at just short of 3.0, now it's past 4.5. Unsure which change in between it is that caused it. Did you observe the same @clinta ?

clinta commented:
@adammmmm, on your poller or your SQL server? Or are they one in the same?

spinza commented:
@clinta I was seeing issues on my server but I've tweaked things a bit and it seems to be running fine now (at least since last night).  I'm still running on the commit 5 days ago.  Will update at some point.

adammmmm commented:
@clinta Yeah, the poller is running the MySQL server.

clinta commented:
@adammmmm, I saw load increase on my pollers with a89018a but it was resolved by removing busy loops in 5e3d968. My MySQL server has not had any change in performance, even switching it to use a single connection hasn't made any difference for me. I'm testing a change that would pull devices in batches for threads to process which should reduce the number expensive queries.

adammmmm commented:
Hmm alright, interestingly enough even the overall traffic on the poller was also increased from around 750Kbit/s to 1.1Mbit/s. 
Screenshot of the graphs: http://imgur.com/7F0uCja

clinta commented:
In syslog, did the number of devices polled per minute increase when you noticed the increased load?

adammmmm commented:
Yes, it looks like it's increased the devices polled per minute.
mysqladmin status shows an increase in queries per second, also opens and open tables have increased somehow.

clinta commented:
Okay, if you're polling more per minute then your poller is working more efficiently and getting more work done, so it's not unexpected that it would use more resources. At that point you can either reduce `$config['poller_service_workers']` or increase `$config['poller_service_poll_frequency']`.

adammmmm commented:
I guess yeah, it's still way less load than with cron jobs, so all good, just figured I'd mention it!

spinza commented:
Updated to 5e3d968 last night and also still running fine.

spinza commented:
@clinta I'm thinking that I need to distribute my mariadb for my setup to scale.  I'm polling 150 devices but I want to poll more like 600+.  Will this work with database replication (galera?) or not?

clinta commented:
@spinza It will work but with caviats for multi-master. If doing multi-master, all pollers in a single poller group must be pointed to the same master server.

spinza commented:
I wanted to give each poller a master db replica and make them fish from the default poller group.  This won't work then?

clinta commented:
@spinza, no because GET_LOCKs don't replicate in Galaria, so none of your pollers would know what the others are doing.

clinta commented:
@spinza, would it be helpful to be able to configure a different SQL server for device queuing and locking? Then you could configure the poller-service to use one SQL server for locking, but poller.php which issues all the writes could go to a more local replica.

spinza commented:
@clinta thanks for the offer, but I think I will need to figure out I guess.  I think I will probably use something closer to poller-wrapper.py (but a service) but still using memcached for locking.  Seems the most convenient given that I have it all still running.

adammmmm commented:
Been two weeks without any issue whatsoever now.

f0o commented:
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Awesome!

So this is good for merge now @clinta ?


On September 23, 2015 8:32:06 AM GMT+02:00, adammmmm <notifications@github.com> wrote:
>Been two weeks without any issue whatsoever now.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/librenms/librenms/pull/1838#issuecomment-142507786
-----BEGIN PGP SIGNATURE-----
Version: APG v1.1.1

iQJPBAEBCgA5BQJWAkfaMhxEYW5pZWwgUHJldXNza2VyIChNb2JpbGVLZXkjMikg
PGYwb0BsaWJyZW5tcy5vcmc+AAoJEJMZKtVWOFzBo+wP/0TvAaCjddr4ElE4n1Yn
Mx0Adc3o+JEzWTuieWNSllYYDBylALnL0z92TFEmOOA0Wq4ErQA+D3SBzLyhKx3k
vx/7DIVAqSZ24Zlyo40gCxYbD8ftr4QDbx5No1bMXySTU9Ts76PhFob1RoxDSpnd
Xz513FsIVQnI3iaHEyo6ZBjmgatjRBd7RxbCx5klN8Yd2H4vxEKQsTbsGXnx+lxY
niPdJY0jZuoT2Nl4myof08Ol4FX0rO8TK6E91wadHLdx0tXwi0lj+9hYUDU6Byfn
vv99820829FLaRNlLl84Jnuy0sYI1Zd2Uy8bgsHdAIBJ2V+wWKbtwk/f0Zh1r5K7
6kxU2PimCSTsyZhf7z/d2PWKQKNGV4fb0TSA+WmIMouGTF9lc0XdCohCvsSw2bcF
qgOfVJsWAWU95JVp6WwH0i9g9Eth7WiqegcNipZC3+FdxIRwYhpdZhVUbJEWoCqf
GN6sT52ipMphm+zO7jChuNA1TsKC1aY9NOeOENw3f7qt9fYUzO+Za5W+ZSSY0MLo
R9WW0TzXLBGwo0OqpC/Lkbkq9fzuMhkF843Uo/w3Lg3ib7zuya2z0bip+eMUQqaj
R0RXMMoIhHgIB/LmZoP5ETm2Y0Gll7Ym1OJdnor49FVwIiSCkEOC0xUbhDDVBpBz
N6DxsL6W581dyCXXqd5eLQbS
=FAsl
-----END PGP SIGNATURE-----

clinta commented:
I've also been running it solid 9 days now. @f0o, anything blocking a merge at this point?

f0o commented:
Great work guys :)
